### DISCUSSION TITLE
Q&A: TaskSampler configuration — strategy selection, curriculum tuning, and difficulty updates

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#264] [Q&A]
## Q&A: TaskSampler configuration and curriculum tuning

Covering the most common configuration questions for `TaskSampler` and `CurriculumScheduler` from Phase 7.2.

---

**Q1: Which `SamplingStrategy` should I start with?**

Start with `UNIFORM` while your task pool is small (< 10 tasks). Once you have 20+ tasks with meaningful difficulty variance, switch to `CURRICULUM`:

| Pool size | Recommended strategy | Reason |
|---|---|---|
| < 10 tasks | `UNIFORM` | Not enough variance to benefit from curriculum |
| 10-50 tasks | `CURRICULUM` | Self-paced learning speeds up early convergence |
| 50+ tasks, diverse | `DIFFICULTY_WEIGHTED` | Focuses compute on genuinely hard tasks |
| Any, forgetting concern | `RECENCY_WEIGHTED` | Prevents easy tasks from being ignored |

---

**Q2: How do I tune `CurriculumScheduler.slope`?**

The slope controls how fast the difficulty ceiling rises. A rough rule of thumb:

| Target | slope setting |
|---|---|
| Reach all tasks in 100 meta-steps | `slope = 0.7 / 100 = 0.007` |
| Reach all tasks in 500 meta-steps | `slope = 0.7 / 500 = 0.0014` |
| Reach all tasks in 1000 meta-steps | `slope = 0.7 / 1000 = 0.0007` |

Formula: `slope = (1.0 - start_ceiling) / target_steps`.

---

**Q3: What happens if `eligible_tasks()` returns an empty set?**

If all registered tasks have difficulty > ceiling (e.g., at the very start of training with `start_ceiling=0.3` and all tasks initialised at `difficulty=0.5`), `sample_batch()` falls back to the single easiest task repeated `batch_size` times, rather than raising an exception. Monitor `asi_task_sampler_eligible_tasks` — if it reads 0, lower `start_ceiling` or pre-initialise task difficulties below the starting ceiling.

---

**Q4: How is `update_difficulty(task_id, loss)` called in practice?**

After each `meta_update()`, iterate over the task batch and the returned `inner_losses`:

```python
result = await meta_learner.meta_update(batch)
for task_sample, loss in zip(batch, result.inner_losses):
    sampler.update_difficulty(task_sample.task_id, loss)
```

The `loss` here is the inner-loop loss (mean squared error or cross-entropy depending on task type). Set `loss_scale` to the expected nominal loss so difficulty stays in [0, 1]:

```python
sampler = TaskSampler(pool, loss_scale=2.5)  # if nominal loss ~ 2.5
```

---

**Q5: Can `TaskSampler` coexist with `TaskRegistry` (Phase 6.5)?**

Yes. `TaskRegistry` manages per-task EWC Fisher matrices; `TaskSampler` manages sampling weights. They share task IDs but don't share state. When `TaskRegistry` evicts a task:

- The evicted task's Fisher matrix is removed from memory
- `TaskSampler` still has a stale difficulty entry for that task
- Call `sampler.deregister_task(task_id)` in the same eviction callback to keep them in sync

---

**Q6: Does `TaskSampler` support `replace=False` sampling (no repeats in a batch)?**

By default, `sample_batch()` samples **with replacement** so the batch can contain the same task multiple times (especially with small pools). Pass `allow_repeats=False` to the constructor to enforce without-replacement sampling. Note: this raises `ValueError` if `batch_size > len(eligible_tasks)`.

---

**Q7: How do I monitor curriculum progress with PromQL?**

```promql
# Current difficulty ceiling
asi_curriculum_ceiling

# Number of tasks currently eligible
asi_task_sampler_eligible_tasks

# Tasks sampled rate (last 5 min, per task)
rate(asi_task_sampler_tasks_sampled_total[5m])

# Alert: eligible tasks dropped to 0 (ceiling too low)
asi_task_sampler_eligible_tasks == 0
```
